libraries/chain/apply_context.cpp

```diff
@@ -74,36 +47,75 @@ action_trace apply_context::exec_one()
       r.auth_sequence[auth.actor] = next_auth_sequence( auth.actor );
    }
 
-   action_trace t(r);
-   t.trx_id = trx_context.id;
-   t.block_num = control.pending_block_state()->block_num;
-   t.block_time = control.pending_block_time();
-   t.producer_block_id = control.pending_producer_block_id();
-   t.account_ram_deltas = std::move( _account_ram_deltas );
-   _account_ram_deltas.clear();
-   t.act = act;
-   t.context_free = context_free;
-   t.console = _pending_console_output.str();
+   trace = {r}; // reset action_trace
+
+   trace.trx_id = trx_context.id;
+   trace.block_num = control.pending_block_state()->block_num;
+   trace.block_time = control.pending_block_time();
+   trace.producer_block_id = control.pending_producer_block_id();
+   trace.act = act;
+   trace.context_free = context_free;
+
+   const auto& cfg = control.get_global_properties().configuration;
+   try {
+      try {
+         const auto& a = control.get_account( receiver );
+         privileged = a.privileged;
+         auto native = control.find_apply_handler( receiver, act.account, act.name );
+         if( native ) {
+            if( trx_context.can_subjectively_fail && control.is_producing_block() ) {
+               control.check_contract_list( receiver );
+               control.check_action_list( act.account, act.name );
+            }
+            (*native)( *this );
+         }
+
+         if( a.code.size() > 0
+             && !(act.account == config::system_account_name && act.name == N( setcode ) &&
+                  receiver == config::system_account_name) ) {
+            if( trx_context.can_subjectively_fail && control.is_producing_block() ) {
+               control.check_contract_list( receiver );
+               control.check_action_list( act.account, act.name );
+            }
+            try {
+               control.get_wasm_interface().apply( a.code_version, a.code, *this );
+            } catch( const wasm_exit& ) {}
+         }
+      } FC_RETHROW_EXCEPTIONS( warn, "pending console output: ${console}", ("console", _pending_console_output.str()) )
+   } catch( fc::exception& e ) {
+      trace.except = e;
+      finalize_trace( trace, start );
+      throw;
+   }
 
    trx_context.executed.emplace_back( move(r) );
 
    if ( control.contracts_console() ) {
-      print_debug(receiver, t);
+      print_debug(receiver, trace);
```

Comment:
`print_debug` will be passed a trace with an empty `console` since it hasn't been set yet with `finalize_trace`. It should be moved to after the `finalize_trace` call below.

Reply:
Moved.